### ISSUE TITLE
Can't sign in to Slack

### ISSUE THREAD

**Issue by jdm** [servo/servo#13218]
https://irccloud.mozilla.com/file/6B4oF3ny/Screen+Shot+2016-09-09+at+09.57.55.png

It says "Unable to set cookie".

**Comment by jdm** [comment]
Relevant output from network:
* the initial 302 response contains a https://slack.com/checkcookie?redir=https%3A%2F%2Frgsoc-jam.slack.com%2F and a Set-Cookie with a `.slack.com` domain
* that cookie is included in the request to checkcookie

**Comment by jdm** [comment]
We're including the pre-request body in the redirected request, which is different than Firefox. However, I'm suspicious that the cookie value seen in Set-Cookie in Servo appears to be percent-encoded, while I don't see that happening in Firefox's network tool.

**Comment by jdm** [comment]
This may be related to https://github.com/alexcrichton/cookie-rs/issues/40.

**Comment by jdm** [comment]
Indeed, applying @retep998's patch to cookie-rs to remove all percent-decoding makes the Slack sign in work:
```diff
diff --git a/src/lib.rs b/src/lib.rs
index 2b8c693..ab4363c 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -30,13 +30,6 @@ pub struct Cookie {
     pub custom: BTreeMap<String, String>,
 }

-fn percent_decode(input: &str) -> Result<String, ()> {
-    match url::percent_encoding::percent_decode(input.as_bytes()).decode_utf8() {
-        Ok(s) => Ok(s.into_owned()),
-        Err(_) => Err(())
-    }
-}
-
 impl Cookie {
     pub fn new(name: String, value: String) -> Cookie {
         Cookie {
@@ -61,11 +54,11 @@ impl Cookie {
         let mut pairs = s.trim().split(';');
         let keyval = match pairs.next() { Some(s) => s, _ => return Err(()) };
         let (name, value) = try!(split(keyval));
-        c.name = try!(percent_decode(name));
+        c.name = name.into();
         if c.name.is_empty() {
             return Err(());
         }
-        c.value = try!(percent_decode(value));
+        c.value = value.into();

         for attr in pairs {
             let (k, v) = attr_split(attr);
@@ -184,10 +177,7 @@ pub struct AttrVal<'a>(pub &'a str, pub &'a str);
 impl<'a> fmt::Display for AttrVal<'a> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         let AttrVal(ref attr, ref val) = *self;
-        write!(f, "{}={}", attr, url::percent_encoding::percent_encode(
-            val.as_bytes(),
-            url::percent_encoding::USERINFO_ENCODE_SET)
-        )
+        write!(f, "{}={}", attr, val)
     }
 }

```
cc @paulrouget

**Comment by jdm** [comment]
Now that https://github.com/alexcrichton/cookie-rs/pull/66 has been merged, we should update our cookie-rs version and see how Slack is doing!

**Comment by jdm** [comment]
This should require running `./mach cargo-update -p cookie` and rebuilding, as well as testing before and after.

**Comment by fkautz** [comment]
@jdm I tried updating the cookie dependency using the command you provided, but the cookie dependency remains at "0.2.5." Any recommendations on how to get this package upgraded? Maybe shared dependencies between this and other projects are preventing the dependency from correctly resolving? e.g. cookie is asking for openssl "^0.9.0" while other deps like hyper require "^0.7"

Any advice?

**Comment by jdm** [comment]
@fkautz You may need to modify the Cargo.toml files that list it to use the new version.

**Comment by fkautz** [comment]
@jdm I was able to use ``[replace]`` and cherry pick the fix to a local branch with some minor modifications.

I was able to reproduce the issue, and verify that the patch fixes this bug.

**Comment by jdm** [comment]
What about using the unmodified cookie-rs from upstream?

**Comment by fkautz** [comment]
"cookie 0.5.1" requires "openssl ^0.9" (OpenSSL 1.0.x) which requires hyper to be updated from "openssl ^0.7" to "openssl ^0.9".

A prior conversation with @jdm pointed me towards #14484 which introduces a new hyper-openssl dependency and disables the optional openssl component in hyper. I had a shot at rebasing it to support the latest cookie serialization changes w/serde, but haven't been successful yet.

**Comment by fkautz** [comment]
In short, #14484 or an equivalent patch needs to be merged before cookie 0.5.1 can be used.

**Comment by jdm** [comment]
This works now.
